lib/spack/spack/variant.py

```diff
@@ -617,6 +623,102 @@ def substitute_abstract_variants(spec):
         spec.variants.substitute(new_variant)
 
 
+class DisjointSetsOfValues(Sequence):
```

Comment:
> Minimally, this would require some updates to the documentation (I need to track down precisely where) - **although I assume you were holding off on that until a conversation was had about the PR**

:+1: 

> While the invocations read off the tongue more smoothly, IMO it is not inherently more comprehensible than a functional approach.

I agree with that for simple variants. For variants with a more complicated logic I think this approach reads better than a functional one and permits to set defaults in a central place (the factory functions like `any_combination_of`). Imo it's also in line with the declarative nature of directives.

> (i.e. as they appear in the context of their prose - what does "with" mean for a builder object? I still have to read the function description or look at the code).

Maybe I don't get your point above, but I don't see where this isn't clear:
```
values=any_combination_of('foo', 'bar').with_default('bar')
```
in the context of the variant directive.

Reply:
> Maybe I don't get your point above, but I don't see where this isn't clear:
> `values=any_combination_of('foo', 'bar').with_default('bar')`
> in the context of the variant directive.

Reading this is fine (in this case), but when a user looks up how to write it, they may be momentarily confused by the paradigm shift. Most questions of "how do I set this or that" boil down to passing the write arguments to the write functions, but in this case it's now a matter of invoking a series of functions; those unfamiliar with it might find it slightly jarring (and to be clear I only think this is worth mentioning because this is exposed to package writers).